drivers/cartesi/cmio.c

```diff
@@ -0,0 +1,231 @@
+////// SPDX-License-Identifier: GPL-2.0
+/*
+ * Cartesi cmio device.
+ * Copyright (C) 2023 Cartesi Machine reference unit
+ */
+
+#include <linux/kernel.h>
+#include <linux/mm.h>
+#include <linux/module.h>
+#include <linux/slab.h>
+#include <linux/ioctl.h>
+#include <linux/uaccess.h>
+#include <linux/fs.h>
+#include <linux/device.h>
+#include <linux/miscdevice.h>
+#include <linux/platform_device.h>
+#include <linux/io.h>
+#include <linux/of.h>
+#include <asm/sbi.h>
+#include <uapi/linux/cartesi/cmio.h>
+
+#define DEVICE_NAME "cmio"
+#define MODULE_DESC "Cartesi Machine " DEVICE_NAME " device"
+
+#define SBI_YIELD 9
+
+struct cmio_device {
+	struct platform_device *pdev;
+	struct miscdevice mdev;
+	struct cmio_setup bufs;
+	atomic_t single_user_lock;
+};
+
+static struct cmio_device *to_cmio_device(struct file *file)
+{
+	struct miscdevice *dev = file->private_data;
+	return container_of(dev, struct cmio_device, mdev);
+}
+
+static long cmio_ioctl_setup(struct cmio_device *me, unsigned long arg)
+{
+	if (copy_to_user((void __user *)arg, &me->bufs, sizeof me->bufs))
+		return -EFAULT;
+
+	return 0;
+}
+
+static long cmio_ioctl_yield(struct cmio_device *me, unsigned long arg)
+{
+	__u64 req = 0,
+	      rep = 0;
+
+	if (copy_from_user(&req, (void __user *)arg, sizeof req))
+		return -EFAULT;
+
+	rep = sbi_ecall(SBI_YIELD, 0, req, 0, 0, 0, 0, 0).value;
+
+	if (copy_to_user((void __user *)arg, &rep, sizeof rep))
+		return -EFAULT;
+
+	return 0;
+}
+
+/*
+ * We enforce only one user at a time here with the open/release.
+ */
+static int cmio_open(struct inode *inode, struct file *file)
+{
+	struct cmio_device *cmio = to_cmio_device(file);
+	if (!cmio)
+		return -EBADF;
+
+	if (!atomic_dec_and_test(&cmio->single_user_lock)) {
```

Comment:
This pattern has precedence in a few places:
- https://elixir.bootlin.com/linux/latest/source/arch/m68k/bvme6000/rtc.c#L137
- https://elixir.bootlin.com/linux/latest/source/drivers/s390/cio/chsc_sch.c#L882
- https://static.lwn.net/images/pdf/LDD3/ch06.pdf (p.174)

But I agree that it looks a bit dodgy (maybe old?). I'll flip the logic around so that only `0` gets the `open` going.

Reply:
The reason this "works" is kind of ugly. `atomic_dec_and_test` only returns true if the *new* value for `single_user_lock` is 0. So the first user to enter the line will return true and initialize the driver. Other users, even preempted, will return false, and `single_user_lock` will be progressively decreased in value over the negative numbers. The only way for another user to manage initializing this is when `single_user_lock` wraps around all the way back to 1. So, in theory, it could happen. In practice, if `int` is big enough, and the scheduler is not insane, this would never happen. So, sorry for the noise.